### PR TITLE
Feature: AI Node and AI Actions extension

### PR DESCRIPTION
**Pull request by ghost**
### Proposed Changes
* Created AI Content node 
* Created AI Actions extension (accept, regenerate, delete content)

Link to the issue: [https://github.com/dotCMS/core/issues/25669](https://github.com/dotCMS/core/issues/25669)

### Checklist
- [ ] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots
Original             |  Updated
:-------------------------:|:-------------------------:
** original screenshot **  |  ** updated screenshot **

### CODE REVIEW

**Comment by fmontes** [comment]
@nikolatrajkovic24 Please clean this PR, no changes outside block editor should be here.

**Review by zJaaal — DISMISSED** [review]

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.html [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<p-listbox
+    [options]="actionOptions"
+    [style]="{
+        width: '12.5rem',
+        padding: '0.5rem',
+        'margin-left': '3.75rem',
+        'background-color': '#fff'
+    }"
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.html:8]: Can we switch to CSS `classes`? That way we can use our CSS variables

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+@Component({
+    selector: 'dot-ai-content-actions',
+    templateUrl: './ai-content-actions.component.html',
+    styleUrls: ['./ai-content-actions.component.scss']
+})
+export class AIContentActionsComponent implements OnInit {
+    @Output() acceptEmitter = new EventEmitter<boolean>();
+    @Output() regenerateEmitter = new EventEmitter<boolean>();
+    @Output() deleteEmitter = new EventEmitter<boolean>();
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.ts:22]: Maybe we can unify these three `@Output` into one called `@Output() actionEmitter` - Naming can be better - and emit the ActionType. Something like:

```ts
enum ACTIONS {
   ACCEPT = "ACCEPT",
   DELETE = "DELETE",
   REGENERATE = "REGENERATE"
}

@Output() actionEmitter = new EventEmitter<ACTIONS>();
```

That way the `@Ouput` will match the `Component Name` keeping consistency.

File: core-web/libs/block-editor/src/lib/nodes/ai-content/ai-content.node.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        };
+    },
+
+    parseHTML() {
+        return [
+            {
+                tag: 'div'
+            }
+        ];
+    },
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/nodes/ai-content/ai-content.node.ts:28]: This function gets the HTML DOM element passed as a parameter. The `div` tag can interfere with other extensions, we should consider changing the selector to something like `div[data-ai-node]`

**Review by rjvelazco — CHANGES_REQUESTED** [review]

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    // delete option
+    &:last-child {
+        border-top: 1px solid $color-palette-gray-300 !important;
+    }
+
+    &:hover {
+        background-color: var(--color-palette-primary-100) !important;
+    }
+}
```
[the remark this answers is not in the document]
Comment (nikolatrajkovic24) [reply core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss:26]: I tried a couple of different ways, but it looks like the selector is weaker and style is not applied unless the !important

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.html (hunk 1 above)
Comment (nikolatrajkovic24) [reply core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.html:8]: Couldnt find another way to apply those styles on this listbox unless this way...

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/plugins/ai-content-actions.plugin.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    private acceptContent() {
+        this.editor.commands.closeAIContentActions();
+        const content = this.component.instance.getLatestContent();
+        this.editor.commands.insertContent(content);
+    }
+
+    private generateContent() {
+        this.editor.commands.closeAIContentActions();
+
+        this.component.instance.getNewContent().subscribe((newContent) => {
```
Comment (fmontes) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-actions/plugins/ai-content-actions.plugin.ts:93]: We need to unsubscribe to cancel the subscription here.

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.html (hunk 1 above)
Comment (fmontes) [reply core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.html:8]: You can pass the property `styleClass` to the `<p-listbox` and then style the class.

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (fmontes) [reply core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss:26]: This is very hard to maintain, not only the important but the child position, we need to find a better way to do this... can you add a screenshot of what you are trying to do here so we can try to find a better.

[the remark this answers is not in the document]
Comment (fmontes) [reply core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss:26]: Also adding `::ng-deep` with selector so generals like `li` and `i` will leak that to other elements, you should add it into the `:host` selector.

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    @Output() actionEmitter = new EventEmitter<ACTIONS>();
+
+    actionOptions!: ActionOption[];
+
+    constructor(private aiContentService: AiContentService) {}
+
+    ngOnInit() {
+        this.actionOptions = [
+            {
+                label: 'Accept',
```
Comment (fmontes) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.ts:35]: What are we with the translation?

File: core-web/libs/block-editor/src/lib/shared/services/ai-content/ai-content.service.ts [hunk 7]
```diff
@@ -13,9 +13,22 @@ interface OpenAIResponse {
 
 @Injectable()
```
Comment (fmontes) [inline core-web/libs/block-editor/src/lib/shared/services/ai-content/ai-content.service.ts:14]: Keep this inside the lib block editor/

File: core-web/libs/block-editor/src/lib/shared/primeng.module.ts [hunk 8]
```diff
@@ -10,6 +10,7 @@ import { InputTextModule } from 'primeng/inputtext';
 import { ListboxModule } from 'primeng/listbox';
 import { MenuModule } from 'primeng/menu';
 import { OrderListModule } from 'primeng/orderlist';
+import { OverlayPanelModule } from 'primeng/overlaypanel';
```
Comment (fmontes) [inline core-web/libs/block-editor/src/lib/shared/primeng.module.ts:13]: Don't modify anything outside the block editor, keep everything in the lib.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/plugins/ai-content-actions.plugin.ts (hunk 5 above)
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-actions/plugins/ai-content-actions.plugin.ts:93]:
```suggestion
        this.component.instance.getNewContent().pipe(takeUntil(this.destroy$)).subscribe((newContent) => {
```

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.html (hunk 1 above)
Comment (rjvelazco) [reply core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.html:8]: See PrimeNg API for [Listbox](https://primeng.org/listbox#api.listbox) Component, we have [styleClass](https://primeng.org/listbox#api.listbox.props.styleClass) and [listStyleClass](https://primeng.org/listbox#api.listbox.props.listStyleClass) Properties.

**Review by rjvelazco — CHANGES_REQUESTED** [review]

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.ts (hunk 6 above)
Comment (nikolatrajkovic24) [reply core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.ts:35]: There is no support for translation implemented yet

File: core-web/libs/block-editor/src/lib/shared/services/ai-content/ai-content.service.ts (hunk 7 above)
Comment (nikolatrajkovic24) [reply core-web/libs/block-editor/src/lib/shared/services/ai-content/ai-content.service.ts:14]: This is part of the block editor- lives under libs/editor/lib/shared you mean on this or something else?

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        // regenerate option
+        &:nth-child(2) {
+            padding: $spacing-3;
+        }
+
+        // delete option
+        &:last-child {
+            border-top: 1px solid $color-palette-gray-300 !important;
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss:26]: Is the `!important` needed?

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        &:hover {
+            background-color: var(--color-palette-primary-100);
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss:30]: Try to use `sass` variables here

```suggestion
            background-color: $color-palette-primary-100;
```

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    background-color: $white;
+
+    // list option
+    .p-listbox-item {
+        background-color: $white;
+        padding: $spacing-2 $spacing-3;
+
+        // accept option
+        &:first-child {
+            background-color: var(--color-palette-primary-200);
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-actions/ai-content-actions.component.scss:16]:
```suggestion
            background-color: $color-palette-primary-200;
```

File: core-web/libs/block-editor/src/lib/extensions/ai-content-actions/plugins/ai-content-actions.plugin.ts [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+        this.component.instance
+            .getNewContent()
+            .pipe(takeUntil(this.destroy$))
+            .subscribe((newContent) => {
+                if (newContent) {
+                    this.editor.commands.deleteSelection();
+                    this.editor.commands.insertAINode(newContent);
+                    this.editor.commands.openAIContentActions();
+                }
+            });
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-actions/plugins/ai-content-actions.plugin.ts:102]: Every time we call this function, we are creating a new subscription. We can handle this in two different ways.

1. Moves this subscription to the `constructor` (Best option - IMO)
2. Use `take(1)` Pipe

```suggestion
        this.component.instance
            .getNewContent()
            .pipe(
              take(1), // Take only the first emission 
              takeUntil(this.destroy$) // If the component is destroyed first
            )
            .subscribe((newContent) => {
                if (newContent) {
                    this.editor.commands.deleteSelection();
                    this.editor.commands.insertAINode(newContent);
                    this.editor.commands.openAIContentActions();
                }
            });
```

File: core-web/libs/block-editor/src/lib/nodes/ai-content/ai-content.node.ts (hunk 3 above)
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/nodes/ai-content/ai-content.node.ts:28]:
```suggestion
    parseHTML() {
        return [
            {
                tag: 'div[ai-content]'
            }
        ];
    },
```

**Review by rjvelazco — CHANGES_REQUESTED** [review]

**Review by fmontes — APPROVED** [review]

**Review by rjvelazco — APPROVED** [review]

**Pull request merged**